packages/components/inputs/rich-text-utils/src/slate-helpers.tsx

```diff
@@ -0,0 +1,240 @@
+      return (
+        <h3 style={style} {...attributes}>
+          {children}
+        </h3>
+      );
+    case BLOCK_TAGS.h4:
+      return (
+        <h4 style={style} {...attributes}>
+          {children}
+        </h4>
+      );
+    case BLOCK_TAGS.h5:
+      return (
+        <h5 style={style} {...attributes}>
+          {children}
+        </h5>
+      );
+    case BLOCK_TAGS.li:
+      return (
+        <li style={style} {...attributes}>
+          {children}
+        </li>
+      );
+    case BLOCK_TAGS.ol:
+      return (
+        <ol style={style} {...attributes}>
+          {children}
+        </ol>
+      );
+    case BLOCK_TAGS.p:
+      return (
+        <p style={style} {...attributes}>
+          {children}
+        </p>
+      );
+    default:
+      return (
+        <p style={style} {...attributes}>
+          {children}
+        </p>
+      );
+  }
+};
+
+const Leaf = ({ attributes, children, leaf }: RenderLeafProps) => {
+  if (leaf.bold) {
+    children = <strong>{children}</strong>;
+  }
+  if (leaf.code) {
+    children = <code>{children}</code>;
+  }
+  if (leaf.italic) {
+    children = <em>{children}</em>;
+  }
+  if (leaf.underline) {
+    children = <u>{children}</u>;
+  }
+  if (leaf.superscript) {
+    children = <sup>{children}</sup>;
+  }
+  if (leaf.subscript) {
+    children = <sub>{children}</sub>;
+  }
+  if (leaf.strikethrough) {
+    children = <del>{children}</del>;
+  }
+
+  return <span {...attributes}>{children}</span>;
+};
+
+const isMarkActive = (editor: TEditor, format: Format) => {
+  const marks = Editor.marks(editor);
+  return marks ? marks[format as keyof typeof marks] === true : false;
+};
+
+const toggleMark = (editor: Editor, format: Format) => {
+  const isActive = isMarkActive(editor, format);
+
+  if (isActive) {
+    Editor.removeMark(editor, format);
+  } else {
+    Editor.addMark(editor, format, true);
+  }
+};
+
+const isBlockActive = (editor: TEditor, format: Format) => {
+  const { selection } = editor;
+  if (!selection) return false;
+
+  const [match] = Array.from(
+    Editor.nodes(editor, {
+      at: Editor.unhangRange(editor, selection),
+      match: (n) =>
+        !Editor.isEditor(n) && SlateElement.isElement(n) && n.type === format,
+    })
+  );
+
+  return Boolean(match);
+};
+
+const toggleBlock = (editor: TEditor, format: Format) => {
+  const isActive = isBlockActive(editor, format);
+  const isList = LIST_TYPES.includes(format);
+
+  Transforms.unwrapNodes(editor, {
+    match: (n) =>
+      !Editor.isEditor(n) &&
+      SlateElement.isElement(n) &&
+      LIST_TYPES.includes(n.type),
+    split: true,
+  });
+  const newProperties: Partial<CustomElement> = {
+    type: isActive ? BLOCK_TAGS.p : isList ? BLOCK_TAGS.li : format,
+  };
+  Transforms.setNodes<SlateElement>(editor, newProperties);
+
+  if (!isActive && isList) {
+    const block: CustomElement = { type: format, children: [] };
+    Transforms.wrapNodes(editor, block);
+  }
+};
```

Comment:
Please mind that that the block and mark toggling logic comes directly from [slate's own implementation of rich text editor](https://github.com/ianstormtaylor/slate/blob/main/site/examples/richtext.tsx#L67:L199)